### ISSUE TITLE
False positive in bugprone-narrowing-conversions?

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#58473]
Hi, 

The following code:

```cpp
#include <chrono>

constexpr float kNano{std::nano::den};
```

Leads to:

```console
warning: narrowing conversion from constant value 1000000000 of type 'intmax_t' (aka 'long') to 'float' [bugprone-narrowing-conversions,cppcoreguidelines-narrowing-conversions]
```

Repro: https://godbolt.org/z/Gacb7o7TE

However `float` can store large numbers up to 10^38, why would it be a problem to store 10^9?

Thanks!

**Comment by JonasToth** [comment]
by definition, the conversion between any floating point type to any integer type is narrowing.

the check could be improved for constants known at compile time, as in your case!
Note that for really large numbers, floating point types can not represent all integers at some point, but only all even ones.
"Any integer with absolute value less than 2^24 can be exactly represented in the single-precision format", see [Wikipedia](https://en.wikipedia.org/wiki/Floating-point_arithmetic#IEEE_754:_floating_point_in_modern_computers) for details.

**Comment by carlosgalvezp** [comment]
The interesting part is that if you replace `nano` with `micro` in my example (1E6 instead of 1E9) then clang-tidy is happy. I'm very curious to understand what's going on :)

**Comment by carlosgalvezp** [comment]
> the conversion between any floating point type to any integer type is narrowing.
Yes, that's correct. However note that the warning is the other way around -> conversion from int to float.

**Comment by carlosgalvezp** [comment]
The threshold seems to be here:

```
constexpr float kNano{16'777'217};
```
Below this number, it's fine, above this number it complains.

**Comment by carlosgalvezp** [comment]
Duh:
https://stackoverflow.com/questions/3793838/which-is-the-first-integer-that-an-ieee-754-float-is-incapable-of-representing-e

> For 32-bit floats, 16,777,217 (224 + 1).
